presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateAggregationsWithDefaultValues.java

```diff
@@ -0,0 +1,148 @@
+
+import java.util.List;
+import java.util.Map;
+
+import static com.facebook.presto.sql.planner.plan.AggregationNode.Step.FINAL;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Scope.REMOTE;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Type.REPARTITION;
+import static com.google.common.base.Preconditions.checkArgument;
+
+public class ValidateAggregationsWithDefaultValues
+        implements Checker
+{
+    private static final SeenExchanges NO_EXCHANGES_SEEN = new SeenExchanges(false, false);
+
+    private final boolean forceSingleNode;
+
+    public ValidateAggregationsWithDefaultValues(boolean forceSingleNode)
+    {
+        this.forceSingleNode = forceSingleNode;
+    }
+
+    @Override
+    public void validate(PlanNode planNode, Session session, Metadata metadata, SqlParser sqlParser, Map<Symbol, Type> types)
+    {
+        planNode.accept(new Visitor(session, metadata, sqlParser, types), null);
+    }
+
+    private class Visitor
+            extends PlanVisitor<SeenExchanges, Void>
+    {
+        final Session session;
+        final Metadata metadata;
+        final SqlParser parser;
+        final Map<Symbol, Type> types;
+
+        public Visitor(Session session, Metadata metadata, SqlParser parser, Map<Symbol, Type> types)
+        {
+            this.session = session;
+            this.metadata = metadata;
+            this.parser = parser;
+            this.types = types;
+        }
+
+        @Override
+        protected SeenExchanges visitPlan(PlanNode node, Void context)
+        {
+            return aggregatedSeenExchanges(node.getSources());
+        }
+
+        @Override
+        public SeenExchanges visitAggregation(AggregationNode node, Void context)
+        {
+            SeenExchanges seenExchanges = aggregatedSeenExchanges(node.getSources());
+
+            if (!node.getStep().equals(FINAL) || !node.hasEmptyGroupingSet()) {
+                return NO_EXCHANGES_SEEN;
+            }
+
+            if (seenExchanges.remoteRepartitionExchange) {
+                // Final aggregation separated from partial by remote repartition exchange.
+                return NO_EXCHANGES_SEEN;
+            }
+
+            // No remote repartition exchange between final and partial aggregation.
+            // Make sure that final aggregation operators are executed on a single node.
+            ActualProperties globalProperties = PropertyDerivations.derivePropertiesRecursively(node, metadata, session, types, parser);
+            checkArgument(forceSingleNode || globalProperties.isSingleNode(),
+                    "Final aggregation with default value not separated from partial aggregation by remote hash exchange");
+
+            if (!seenExchanges.localRepartitionExchange) {
+                // No local repartition exchange between final and partial aggregation.
+                // Make sure that final aggregation operators are executed by single thread.
```

Comment:
You meant `Make sure that final aggregation operators are NOT executed by single thread`
Right?

Reply:
They must be executed in a single thread. Otherwise default values will be duplicated.